File: ml/game/ipc/channel.lua
-------------------------------------------------------------------------
--	ml/game/ipc/channel.lua -- file transport (game side).
--
--	One-writer-per-file law: the game writes hello/obs/ack, the host
--	writes act. Nothing is written by both sides, ever.
--
--	Write atomicity (engine reality): os.rename cannot replace an
--	existing file on Windows, so every game-side write is tmp -> remove
--	-> rename with a brief NO-FILE window. That gap is part of the
--	protocol: the host reader serves last-good on ENOENT (codified in
--	host/channel.py -- v1's cockpit blanked instead; fixed contract).
--
--	Read side: the act file is read whole, seq-gated (small monotonic
--	ints only -- float32 Lua numbers), battle_id-gated, and DELETED at
--	battle start so a previous session's final order can never replay
--	(v1 harness bug, fixed here by design).
-------------------------------------------------------------------------

local util = require "ml/util";

local M = {};

M.PATH = {
	hello = "data/aai_ml_hello.json",	-- game -> host, once per battle
	obs = "data/aai_ml_obs.json",		-- game -> host, per decision tick
	ack = "data/aai_ml_ack.json",		-- game -> host, act results ring
	act = "data/aai_ml_act.txt",		-- host -> game, atomic (os.replace)
};

local last_seq = -1;

-- Wipe inbound state for a new battle: delete any stale act file, reset
-- the seq gate to -1. Also blanks ack with an id-only stamp (so the host
-- reads "nothing applied yet", not the previous battle's ring).
-- battle_id is our own os.date stamp -- safe to splice into the literal.
function M.reset(battle_id)
	pcall(os.remove, M.PATH.act);
	last_seq = -1;
	M.write(M.PATH.ack,
		'{"battle_id":"' .. tostring(battle_id) .. '","last":null,"ring":[]}');
end;

-- Game-side file write: tmp -> remove -> rename (the documented gap).
function M.write(path, text)
	local tmp = path .. ".tmp";
	local f = io.open(tmp, "wb");
	if not f then return false; end;
	f:write(text);
	f:close();
	os.remove(path);
	local ok = os.rename(tmp, path);
	return ok and true or false;
end;

-- Poll the act file: returns raw text of a NEW frame exactly once
-- (seq > last_seq and battle_id matches), else nil. Caller parses; a
-- parse error still consumes the seq (ERR-acked, never re-parsed).
function M.poll_act(battle_id)
	return nil, "todo(read whole file pcall'd; extract seq+id from line 1 cheaply; gate; advance last_seq)";
end;

return M;

File: ml/host/channel.py
"""ml/host/channel.py -- file transport (host side).

The host half of the IPC contract (game half: ml/game/ipc/channel.lua):

- Reads game-written JSON (hello/obs/ack). Game-side Lua CANNOT replace
  a file atomically on Windows (tmp -> remove -> rename has a no-file
  window), so the reader contract is LAST-GOOD: on ENOENT *or* a
  truncated/mid-write parse failure, serve the cached copy with its old
  mtime-age. (v1's cockpit returned None on ENOENT and blanked the UI --
  that bug is fixed here by contract.)
- Writes the act file ATOMICALLY: tmp + os.replace, explicit utf-8.
  The game side may therefore parse any act file it sees as complete.
- Staleness is mtime-based (the game freezes on pause -- a stale obs is
  normal, never an error); freshness of CONTENT is the seq field.
- Seqs are small monotonic ints (game Lua numbers are float32); the act
  seq counter reseeds from the act file left on disk across restarts.
"""

from __future__ import annotations

import os

GAME_DATA = r"C:\Program Files (x86)\Steam\steamapps\common\Total War Attila\data"

PATH = {
    "hello": "aai_ml_hello.json",  # game -> host, once per battle
    "obs": "aai_ml_obs.json",      # game -> host, per decision tick
    "ack": "aai_ml_ack.json",      # game -> host, apply-results ring
    "act": "aai_ml_act.txt",       # host -> game, atomic
}


def find_game_dir() -> str:
    """The game data dir: cwd/data if it holds THIS tree's log
    (attila_ml_log.txt -- not v1's attila_ai_log.txt), else the Steam path.
    """
    cwd_data = os.path.join(os.getcwd(), "data")
    if os.path.isfile(os.path.join(cwd_data, "attila_ml_log.txt")):
        return cwd_data
    return GAME_DATA


class Channel:
    """One battle-agnostic transport over the game data dir."""

    def __init__(self, data_dir: str | None = None):
        self.data_dir = data_dir or find_game_dir()
        self._cache: dict = {}  # path -> (parsed, mtime)
        self._act_seq: int | None = None  # reseeded lazily from disk

    def read_json(self, name: str):
        """(payload, age_seconds) with the LAST-GOOD contract above.

        (None, None) only before the first successful read ever.
        """
        raise NotImplementedError("todo: mtime cache; ENOENT/parse-fail -> cached copy")

    def poll_obs(self, last_seq: int):
        """A NEW obs frame exactly once: payload iff seq > last_seq, else None."""
        raise NotImplementedError("todo: read_json(obs); gate on payload['seq']")

    def write_act(self, text: str) -> None:
        """Atomic write of one act frame (tmp + os.replace, utf-8)."""
        raise NotImplementedError("todo: tmp write + os.replace")

    def next_act_seq(self) -> int:
        """Next small monotonic act seq (reseed from the on-disk act file once)."""
        raise NotImplementedError("todo: parse existing act frame line for seq on first call")
